pkg/scheduler/scheduler.go

```diff
@@ -429,20 +458,275 @@ func (sched *Scheduler) bind(assumed *v1.Pod, b *v1.Binding) error {
+				if err != nil {
+					klog.V(3).Infof("Instance Delete Failed!")
+				}
+				go sched.startScheduling(scheduleResultQueue, pod, tokenMap, manifest)
+				break
+			}
+			time.Sleep(2 * time.Second)
+			instanceStatus, err = checkInstanceStatus(host, authToken, instanceID)
+			if err != nil {
+				return
+			} else {
+				continue
+			}
+		} else if instanceStatus == "ERROR" {
+			klog.V(3).Infof("Instance Status: %v", instanceStatus)
+			// Send delete instance request
+			err := deleteInstance(host, authToken, instanceID)
+			if err != nil {
+				klog.V(3).Infof("Instance Delete Failed!")
+			}
+			go sched.startScheduling(scheduleResultQueue, pod, tokenMap, manifest)
+		} else if instanceStatus == "ACTIVE" {
+			klog.V(3).Infof("Instance Status: %v", instanceStatus)
+			sched.config.PodPhaseUpdater.Update(pod, v1.PodRunning)
+		}
+		break
+	}
+}
+
+func scheduleResultEnqueue(scheduleResultQueue *internalqueue.Queue, scheduleResult core.ScheduleResult) {
+	scheduleResultQueue.Put(scheduleResult.SuggestedHost)
+}
+
+func (sched *Scheduler) scheduleResultDequeue(scheduleResultQueue *internalqueue.Queue, tokenMap map[string]string, manifest *v1.PodSpec, pod *v1.Pod) {
+	res, _ := scheduleResultQueue.Get(1)
+	host := fmt.Sprintf("%v", res[0])
+
+	authToken, exist := tokenMap[host]
+	if !exist || tokenExpired(host, authToken) {
+		// Post Request a new token
+		newToken, err := requestToken(host)
+		if err != nil {
+			klog.V(3).Infof("Token Request Failed.")
+			return
+		}
+		authToken = newToken
+		// Update tokenMap
+		tokenMap[host] = authToken
+	}
+
+	instanceID, err := serverCreate(host, authToken, manifest)
+	if err != nil {
+		klog.V(3).Infof("Server create failed")
+		return
+	}
+	klog.V(3).Infof("Instance ID: %v", instanceID)
+
+	go sched.scheduleInstanceStatusCheck(host, authToken, instanceID, pod, scheduleResultQueue, manifest)
+}
+
+func (sched *Scheduler) startScheduling(scheduleResultQueue *internalqueue.Queue, pod *v1.Pod, tokenMap map[string]string, manifest *v1.PodSpec) {
+	scheduleResult, _ := sched.globalSchedule(pod)
+	go scheduleResultEnqueue(scheduleResultQueue, scheduleResult)
+	go sched.scheduleResultDequeue(scheduleResultQueue, tokenMap, manifest, pod)
+}
+
+func (sched *Scheduler) globalScheduleOne() {
+	// 1. Get the Pod to be scheduled from the queue
+	pod := sched.config.NextPod()
+	// pod could be nil when schedulerQueue is closed
+	if pod == nil || pod.Status.Phase == v1.PodRunning {
+		return
+	}
+	if pod.DeletionTimestamp != nil {
+		sched.config.Recorder.Eventf(pod, v1.EventTypeWarning, "FailedScheduling", "skip schedule deleting pod: %v/%v/%v", pod.Tenant, pod.Namespace, pod.Name)
+		klog.V(3).Infof("Skip schedule deleting pod: %v/%v/%v", pod.Tenant, pod.Namespace, pod.Name)
+		return
+	}
+
+	manifest := &(pod.Spec)
+
+	klog.V(3).Infof("Attempting to schedule pod: %v/%v/%v", pod.Tenant, pod.Namespace, pod.Name)
+
+	go sched.startScheduling(scheduleResultQueue, pod, tokenMap, manifest)
+}
```

Comment:
Hi Kevin, as we discussed, we should not have one thread per POD request. Also we need to start a timer before starting the scheudling.